engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java

```diff
@@ -77,29 +79,32 @@ public long activateElement(
    * @param processInstanceRecord the record of the process instance
    * @param elementToActivate The element to activate
    * @param createVariablesCallback Callback to create variables at a given scope
-   * @return The key of the activated element instance
+   * @return The key of the activated element instance and the keys of all it's flow scopes
    */
-  public long activateElement(
+  public ActivatedElementKeys activateElement(
       final ProcessInstanceRecord processInstanceRecord,
       final AbstractFlowElement elementToActivate,
       final BiConsumer<DirectBuffer, Long> createVariablesCallback) {
-    final var flowScopes = collectFlowScopesOfElement(elementToActivate);
+    final var activatedElementKeys = new ActivatedElementKeys();
```

Comment:
🔧 Currently, this object is created here and then passed into the `activateFlowScopes` method as a parameter, which means it's actively modified. I think we can make the ActivatedElementKeys class immutable by constructing the instance inside the `activateFlowScopes` method on demand. I think that this would allow changing the ActivatedElementKeys into a record as well.

Reply:
I don't think this is easily possible. The `activateFlowScopes` method is recursive. This would mean the objects gets recreated with each call. I'd have to recreate it with a list of flow scope of the previous + the current one.
Besides that I won't have the element instance key of the element that's currently getting activated available yet.

There is probably a way to do it, but I believe the constant recreation of the object will make the code less understandable.